### PR TITLE
MERGE READY: Cleanup Merge of Death

### PR DESCRIPTION
**Pull request by devshgraphicsprogramming**
Almost everything works (not BRDF Explorer)

### CODE REVIEW

File: src/irr/asset/IAssetManager.cpp [hunk 1]
```diff
@@ -2,36 +2,155 @@
...
+
+#ifdef _IRR_COMPILE_WITH_BMP_WRITER_
+#include "irr/asset/CImageWriterBMP.h"
+#endif
+
 using namespace irr;
 using namespace asset;
 
-std::function<void(IAsset*)> irr::asset::makeAssetGreetFunc(const IAssetManager* const _mgr)
+std::function<void(SAssetBundle&)> irr::asset::makeAssetGreetFunc(const IAssetManager* const _mgr)
```
Comment (devshgraphicsprogramming) [inline src/irr/asset/IAssetManager.cpp:69]: @Crisspl check if the asset writers on your branch with non-ptr caches still work!

File: src/irr/asset/CImageLoaderJPG.cpp [hunk 2]
```diff
@@ -285,19 +278,20 @@ asset::SAssetBundle CImageLoaderJPG::loadAsset(io::IReadFile* _file, const asset
...
+	constexpr uint32_t MaxJPEGResolution = 65535u;
+	uint8_t* rowPtr[MaxJPEGResolution];
 	for (uint32_t i = 0; i < height; ++i)
 		rowPtr[i] = &reinterpret_cast<uint8_t*>(output->getPointer())[i*rowspan];
 
 	// Read rows from bottom order to match OpenGL coords
 	uint32_t rowsRead = 0;
 	while (cinfo.output_scanline < cinfo.output_height)
-		rowsRead += jpeg_read_scanlines(&cinfo, &rowPtr[cinfo.output_height - 1 - rowsRead], 1);
+		rowsRead += jpeg_read_scanlines(&cinfo, &rowPtr[rowsRead], 1);
```
Comment (devshgraphicsprogramming) [inline src/irr/asset/CImageLoaderJPG.cpp:294]: @Crisspl that was our problem XD

Comment (Crisspl) [reply src/irr/asset/CImageLoaderJPG.cpp:294]: yea there was some reason why someone did this but I don't remember

Comment (devshgraphicsprogramming) [reply src/irr/asset/CImageLoaderJPG.cpp:294]: after you get back BRDF Explorer needs a fix

File: include/IDriver.h [hunk 3]
```diff
@@ -278,7 +278,7 @@ namespace video
             }
 
             template<typename AssetType>
-            core::vector<core::smart_refctd_ptr<typename video::asset_traits<AssetType>::GPUObjectType> > getGPUObjectsFromAssets(AssetType* const* const _begin, AssetType* const* const _end, IGPUObjectFromAssetConverter* _converter = nullptr);
+            created_gpu_object_array<AssetType> getGPUObjectsFromAssets(AssetType* const* const _begin, AssetType* const* const _end, IGPUObjectFromAssetConverter* _converter = nullptr);
```
Comment (devshgraphicsprogramming) [inline include/IDriver.h:281]: @Crisspl familiarize yourself with the changes in this commit

**Comment by devshgraphicsprogramming** [comment]
@Crisspl do you mind removing the derivative mapping from example 15? It makes the example waay to convoluted and complex (it already has skinning, geometry shader and shadows).

Also less stuff to maintain

**Comment by devshgraphicsprogramming** [comment]
@Crisspl now you need to "update" & "fix" BRDF Explorer for the new `smart_refctd_ptr` changes and the texture Y-flip on models

**Comment by devshgraphicsprogramming** [comment]
@Crisspl the only thing thats absolutely fucked right now is the normal quantization in MSVC Release builds... have to find out why, other than that, no bugs

File: include/IDriver.h (hunk 3 above)
Comment (Crisspl) [reply include/IDriver.h:281]: which commit is this actually?

Comment (Crisspl) [reply include/IDriver.h:281]: github doesn't show me

Comment (devshgraphicsprogramming) [reply include/IDriver.h:281]: https://github.com/Crisspl/IrrlichtBAW/pull/10/commits/2c8f3150edcaa5faa7c7fd957ef871fb3e3b164c

**Comment by devshgraphicsprogramming** [comment]
@Crisspl btw, when fixing BRDF Explorer, see if we don't need to update some 3rdparty submodules, especially glew, because I've updated the `glext.h` etc. for OpenGL 4.6+

**Comment by devshgraphicsprogramming** [comment]
Ok I killed the last remaining bug!

**Comment by Crisspl** [comment]
oh nice, congratz, this must have been a tough one

**Comment by devshgraphicsprogramming** [comment]
FUCK YEAH!

**Comment by Crisspl** [comment]
XDD

**Comment by Crisspl** [comment]
nom im going to merge shaders branch with master

**Pull request merged**